src/blog/wasm-decompile.md

```diff
@@ -0,0 +1,233 @@
+---
+title: "What's in that .wasm? Introducing: wasm-decompile"
+author: 'Wouter van Oortmerssen'
+avatars:
+  - 'wouter-van-oortmerssen'
+date: 2020-04-20 12:35:00
+tags:
+  - WebAssembly
+  - tooling
+description: 'WABT gains a new decompilation tool that can make it easier to
+read the contents of Wasm modules'
+tweet: '?'
+---
+We have a growing number of compilers and other tools that generate or
+manipulate `.wasm` files, and sometimes you might want to have a look inside.
+Maybe you're a developer of such a tool, or more directly, you're simply a
+programmer targeting Wasm, and wondering what the generated code looks like,
+for performance or other reasons.
+
+Problem is, Wasm is rather low level, much like actual assembly code. In particular,
+unlike say, the JVM, all data structures have been compiled down to load/store
+operations, rather than conveniently named classes and fields. Compilers like
+LLVM can do an impressive amount of transformations that make the generated code
+look nothing like the code that went in.
+
+Disassemble or.. decompile?
+---------------------------
+
+You could use tools like `wasm2wat` (part of the
+[WABT](https://github.com/WebAssembly/wabt) toolkit), to transform a `.wasm`
+into Wasm's standard text format, `.wat`, which is a very faithful but not
+particularly readable representation.
+
+For example, a simple C function like a dot product:
+
+```C
+typedef struct { float x, y, z; } vec3;
+
+float dot(const vec3 *a, const vec3 *b) {
+    return a->x * b->x +
+           a->y * b->y +
+           a->z * b->z;
+}
+```
+
+We use `clang dot.c -c -target wasm32 -O2` followed by `wasm2wat dot.o` to
+turn it into this `.wat`:
+
+```lisp
+(func $dot (type 0) (param i32 i32) (result f32)
+  local.get 0
+  f32.load
+  local.get 1
+  f32.load
+  f32.mul
+  local.get 0
+  f32.load offset=4
+  local.get 1
+  f32.load offset=4
+  f32.mul
+  f32.add
+  local.get 0
+  f32.load offset=8
+  local.get 1
+  f32.load offset=8
+  f32.mul
+  f32.add))
+```
+
+That is a tiny bit of code, but already not great to read for many reasons.
+Besides the lack of an expression based syntax and general verbosity,
```

Comment:
To be completely fair, it might be helpful to compare with a folded WAT (`wasm2wat -f`) which makes it somewhat more readable.

Reply:
good point, will fix.